### DISCUSSION TITLE
per-file License headers

### DISCUSSION THREAD

**Discussion by fwc** [nebulastream/nebulastream#288] [General]
Looking around how other projects do their per-file license header, I saw:

- the Linux kernel uses [SPDX-License-Identifiers](https://spdx.dev/learn/handling-license-info/), e.g. files start with `// SPDX-License-Identifier: GPL-2.0`
- ClickHouse uses _no_ per-file header at all, they only have the `LICENSE` file in the repo.

---

To use the Apache License, [Apache says](https://apache.org/foundation/license-faq.html#Apply-My-Software) to:

> Include a copy of the Apache License, typically in a file called LICENSE, in your work, and consider also including a NOTICE file.
> 
> It is also valuable to tag each of your source-code files in case they become detached from the LICENSE file. To apply the Apache License to your source-code files, one approach is to attach the following notice to as a comment at the top of each file. Replace the copyright templates with your own identifying information:
> 
> ```
> Copyright [yyyy] [name of copyright owner]
> 
> Licensed under the Apache License, Version 2.0 (the "License");
> you may not use this file except in compliance with the License.
> You may obtain a copy of the License at
> 
>     https://www.apache.org/licenses/LICENSE-2.0
> 
> Unless required by applicable law or agreed to in writing, software
> distributed under the License is distributed on an "AS IS" BASIS,
> WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> See the License for the specific language governing permissions and
> limitations under the License.
> ```
> 
> A shorter variant you may wish to use is:
> 
> ```
> Copyright [yyyy] [name of copyright owner]
> SPDX-License-Identifier: Apache-2.0
> ```

---

The github-backed [choosealicense.com](https://choosealicense.com/licenses/apache-2.0/) mentions the per-file header as an optional step for adding the Apache 2.0 License.

---

Currently, we are missing the "identifying information", i.e. `Copyright [yyyy] [name of copyright owner]` to fully compily with the "also valuable if detached" recommendation.

So what about

- adding the missing `Copyright ...` line? (is DIMA the copyright owner?)
- Switching to the shorter SPDX format?
- removing the (optional but recommended) per-file license header alltogether?

**Comment by adrianmichalke** [comment]
I like the short SPDX format.

[TU Munich's LingoDB](https://github.com/lingo-db/lingo-db/blob/main/LICENSE) just mention:
 `Copyright 2021 LingoDB Contributors (see https://github.com/lingo-db/lingo-db/graphs/contributors)`
[Peloton (CMU)](https://github.com/cmu-db/peloton/blob/master/src/brain/cluster.cpp): 
`Copyright (c) 2015-2018, Carnegie Mellon University Database Group`

**Reply by alepping** [reply]
My preferred option would be to have a single LICENSE file, instead of a per-file header. If we want to keep the per-file header, I'd also advocate for SPDX.

**Comment by alepping** [comment]
It looks like DuckDB also has a central LICENSE file.
In contrast, Materialize has a header in each file: https://github.com/MaterializeInc/materialize/blob/main/src/adapter/src/flags.rs

**Comment by PhilippGrulich** [comment]
I think that this depends on the license. As far as I know, Apache requires per-file license headers or at least its recommended.

**Reply by alepping** [reply]
As long as we are not open source, do we actually need the license? Is it not something that could be easily patched as soon as we go open source? Depending on whether we become an Apache project then, we could add a license to each header.
